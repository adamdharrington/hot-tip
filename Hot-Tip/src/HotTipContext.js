import {createContext} from 'react'

export const HotTipContext = createContext([{}, e => e])
export const HotTipConsumer = HotTipContext.Consumer
