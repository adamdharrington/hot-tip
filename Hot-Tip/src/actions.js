import * as types from './action-types'

export function addTooltip(tip, position) {
  return {
    type: types.UPDATE_TOOLTIP,
    payload: {
      tip,
      visible: true,
      position,
    },
  }
}

export function clearTooltip() {
  return {
    type: types.UPDATE_TOOLTIP,
    payload: {
      visible: false,
    },
  }
}
