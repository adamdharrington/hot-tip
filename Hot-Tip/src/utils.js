export const getBottom = (bounds, force) => {
  const {left, top, width, height, clientWidth, clientHeight} = bounds

  if (!force && top + height + 40 >= clientHeight) {
    return getTop(bounds, true) // eslint-disable-line
  }
  const x = left + width / 2
  const y = top + height
  const leftOffset = x < 100
  const rightOffset = !leftOffset && x + 100 > clientWidth
  const layoutClass = leftOffset
    ? 'left-offset'
    : rightOffset
    ? 'right-offset'
    : 'centred' // eslint-disable-line

  return {
    x,
    y,
    positionClass: 'bottom',
    layoutClass,
  }
}

export const getTop = (bounds, force) => {
  const {top} = bounds
  const bottom = getBottom(bounds, true)

  if (!force && top - 40 <= 0) {
    return bottom
  }

  return {...bottom, y: top, positionClass: 'top'}
}

export const getRight = (bounds, force) => {
  const {left, top, width, height, clientWidth, clientHeight} = bounds

  if (!force && left + width + 200 >= clientWidth) {
    return getLeft(bounds, true) // eslint-disable-line
  }
  const x = left + width
  const y = top + height / 2
  const topOffset = y < 50
  const bottomOffset = !topOffset && y + 50 > clientHeight
  const layoutClass = topOffset
    ? 'top-offset'
    : bottomOffset
    ? 'bottom-offset'
    : 'mid' // eslint-disable-line

  return {
    x,
    y,
    positionClass: 'right',
    layoutClass,
  }
}

export const getLeft = (bounds, force) => {
  const {left} = bounds
  const right = getRight(bounds, force)

  if (!force && left - 200 <= 0) {
    return right
  }

  return {...right, x: left, positionClass: 'left'}
}

export const getPosition = position => {
  switch (position) {
    case 'top':
      return getTop
    case 'right':
      return getRight
    case 'left':
      return getLeft
    case 'bottom':
    default:
      return getBottom
  }
}

export const getBounds = e => {
  const {top, left, width, height} = e.currentTarget.getBoundingClientRect()
  const {pageXOffset, pageYOffset} = window
  const {
    scrollTop,
    clientTop,
    scrollLeft,
    clientLeft,
    clientWidth,
    clientHeight,
  } = document.body
  const absTop = top + Math.max(pageYOffset, scrollTop) - clientTop
  const absLeft = left + Math.max(pageXOffset, scrollLeft) - clientLeft

  return {
    top: absTop,
    left: absLeft,
    width,
    height,
    clientWidth,
    clientHeight,
  }
}

export default {
  getBounds,
  getPosition,
}
