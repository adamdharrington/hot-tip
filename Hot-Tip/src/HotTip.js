import {useContext, useEffect, createElement, cloneElement} from 'react'
import PropTypes from 'prop-types'
import {HotTipContext} from './HotTipContext'
import {clearTooltip, addTooltip} from './actions'
import {getPosition, getBounds, getRenderType} from './utils'

export default function HotTip({position, tip, children}) {
  const dispatch = useContext(HotTipContext)[1]
  useEffect(() => hide, [])

  function show(e) {
    const bounds = getBounds(e)
    const actualPosition = getPosition(position)(bounds)

    dispatch(addTooltip(tip, actualPosition))
  }
  function hide() {
    dispatch(clearTooltip())
  }
  const renderType = getRenderType(children)

  const hotTipProps = {
    tabIndex: 0,
    'aria-describedby': 'ht-anchor',
    onMouseOver: show,
    onFocus: show,
    onMouseOut: hide,
    onBlur: hide,
  }

  if (tip) {
    if (renderType === 'WRAP') {
      return createElement('span', hotTipProps, children)
    }
    if (renderType === 'RENDER_PROPS') {
      return children(hotTipProps)
    }
    return cloneElement(children, hotTipProps)
  }
  return children
}

HotTip.defaultProps = {
  position: 'bottom',
  tip: '',
}

HotTip.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  position: PropTypes.oneOf(['top', 'bottom', 'left', 'right']).isRequired,
  tip: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
}
