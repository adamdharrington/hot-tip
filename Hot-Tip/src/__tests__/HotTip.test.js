/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react'
import renderer, {act} from 'react-test-renderer'
import HotTip from '../HotTip'
import {clearTooltip} from '../actions'
import {HotTipContext} from '../HotTipContext'
import * as utils from '../utils'

describe('HotTip', () => {
  it('renders', () => {
    const component = renderer.create(
      <HotTip tip="click me">
        <a>{'link'}</a>
      </HotTip>,
    )
    const tip = component.toJSON()

    expect(tip.type).toEqual('a')
    expect(tip.children[0]).toEqual('link')
    expect(tip).toMatchSnapshot()
  })

  it('calls render props', () => {
    const renderPropFn = jest.fn(props => {
      return <h2 {...props}>hi</h2>
    })
    const component = renderer.create(
      <HotTip tip="click me">{renderPropFn}</HotTip>,
    )
    const tip = component.toJSON()
    expect(renderPropFn).toHaveBeenCalledTimes(1)
    expect(Object.keys(renderPropFn.mock.calls[0][0]).sort()).toEqual([
      'aria-describedby',
      'onBlur',
      'onFocus',
      'onMouseOut',
      'onMouseOver',
      'tabIndex',
    ])
    expect(tip.type).toEqual('h2')
    expect(tip).toMatchSnapshot()
  })

  it('renders without tip', () => {
    HotTip.PropTypes = {}
    const component = renderer.create(<HotTip>text</HotTip>)
    const tip = component.toJSON()

    expect(tip).toEqual('text')
    expect(tip).toMatchSnapshot()
  })

  it('calls context on hover & blur', () => {
    /* eslint-disable import/namespace */
    utils.getBounds = jest.fn(one => ({one}))
    utils.getPosition = jest.fn(one => two => ({one, two}))
    const dispatch = jest.fn(action => action)
    const component = renderer.create(
      <HotTipContext.Provider value={[{}, dispatch]}>
        <HotTip tip="i am text" position="top">
          text
        </HotTip>
      </HotTipContext.Provider>,
    )
    let tip = component.toJSON()

    act(() => {
      // manually trigger the callback
      tip.props.onMouseOver('hello')
    })
    expect(utils.getBounds).toBeCalledWith('hello')
    expect(utils.getPosition).toBeCalledWith('top')

    tip = component.toJSON()
    expect(tip).toMatchSnapshot()

    act(() => {
      tip.props.onMouseOut()
    })
    tip = component.toJSON()
    expect(tip).toMatchSnapshot()

    const actions = dispatch.mock.calls

    expect(actions[0][0].payload.position).toEqual({
      one: 'top',
      two: {one: 'hello'},
    })
    expect(actions[1][0].payload).toEqual({visible: false})
  })

  it('unmounts calling clear', () => {
    const dispatch = jest.fn(action => action)
    const component = renderer.create(
      <HotTipContext.Provider value={[{}, dispatch]}>
        <HotTip tip="click me">
          <a>{'link'}</a>
        </HotTip>
      </HotTipContext.Provider>,
    )

    act(() => {
      component.unmount()
    })
    expect(dispatch.mock.calls[0][0]).toEqual(clearTooltip())
  })
})
