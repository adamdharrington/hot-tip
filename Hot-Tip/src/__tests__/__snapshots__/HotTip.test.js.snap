// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotTip calls context on hover & blur 1`] = `
<span
  aria-describedby="ht-anchor"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  tabIndex={0}
>
  text
</span>
`;

exports[`HotTip calls context on hover & blur 2`] = `
<span
  aria-describedby="ht-anchor"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  tabIndex={0}
>
  text
</span>
`;

exports[`HotTip calls render props 1`] = `
<h2
  aria-describedby="ht-anchor"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  tabIndex={0}
>
  'hi'
</h2>
`;

exports[`HotTip renders 1`] = `
<a
  aria-describedby="ht-anchor"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  tabIndex={0}
>
  link
</a>
`;

exports[`HotTip renders without tip 1`] = `"text"`;
